languages/bash.scm

```diff
@@ -23,13 +32,88 @@
   (if_statement)
   (list)
   (pipeline)
+  (redirected_statement)
   (subshell)
   (variable_assignment)
   (while_statement)
-  ; TODO: etc.
 ] @allow_blank_line_before
 
-; Surround with spaces
+; Insert a new line after multi-line syntactic blocks or, for where
+; single-line variants exists, after the "closing" subnodes (the
+; specificity is to avoid targeting the single-line context)
+[
+  (if_statement)
+  (case_statement)
+  (do_group)
+] @append_hardline
+
+(subshell
+  ")" @append_empty_softline
+  .
+)
+
+(compound_statement
+  "}" @append_empty_softline
+  .
+)
+
+; A run of "units of execution" (see below, sans variables which are
+; special) and function definitions should be followed by a new line,
+; before a multi-line syntactic block or variable.
+(
+  [
+    (command)
+    (compound_statement)
+    (function_definition)
+    (list)
+    (pipeline)
+    (redirected_statement)
+    (subshell)
+  ] @append_empty_softline
+  .
+  [
+    (c_style_for_statement)
+    (case_statement)
+    (comment)
+    (compound_statement)
+    (declaration_command)
+    (for_statement)
+    (function_definition)
+    (if_statement)
+    (subshell)
+    (variable_assignment)
+    (while_statement)
+  ]
+)
+
+; A run of variable declarations and assignments should be followed by a
+; new line, before anything else
+(
+  [
+    (declaration_command)
+    (variable_assignment)
+  ] @append_hardline
+  .
+  [
+    (c_style_for_statement)
+    (case_statement)
+    (command)
+    (comment)
+    (compound_statement)
+    (compound_statement)
+    (for_statement)
+    (function_definition)
+    (if_statement)
+    (list)
+    (pipeline)
+    (redirected_statement)
+    (subshell)
+    (subshell)
+    (while_statement)
+  ]
```

Comment:
Do you need this? Shouldn't the above always be followed by a hardline?

Reply:
Variable declarations and assignments get a `@prepend_hardline` elsewhere, so they appear on a new line that way. This block also adds a hardline after a run of variables followed by one of the nodes in that alternation, which are also `@prepend_hardline`. This formats as a blank link between a run of variables and those nodes.

If I comment out this anchor + alternation, it has two effects:
1. Every variable declaration and assignment is separated by blank lines (redundant, but syntactically OK)
2. There are cases where declarations and assignments can be mixed and this will split them (which is syntactically invalid). For example:

    ```bash
    export A B=1 C
    ```

    ...gets formatted as:

    ```bash
    export A B=1
    C
    ```

So this somewhat complicated rule is largely from my choice of prepending hardlines, rather than appending them (see #186; specifically the commit message in 968852323). The choice of prepending was because I couldn't get "indent blocks" to work correctly using an append strategy without either an errant new line or spurious indentation:

1. ```bash
   while true; do
     foo

   done
   ```

2. ```bash
   while true; do
     foo
     done
   ```

However, my hunch is that this choice is what's making comments particularly difficult (impossible?) to format; hence #186. (It's also "weird" in the sense that it breaks expectations, which is also something worth correcting.)